File: cmd/14-tcpstate/tcpstate.c
//go:build ignore

#include "vmlinux.h"
#include <bpf/bpf_tracing.h>
#include <bpf/bpf_helpers.h>
#include <bpf/bpf_core_read.h>

#define MAX_ENTRIES 10240
#define AF_INET 2
#define AF_INET6 10
#define TASK_COMM_LEN 16

struct event
{
    unsigned __int128 saddr;
    unsigned __int128 daddr;
    __u64 skaddr;
    __u64 ts_us;
    __u64 delta_us;
    __u32 pid;
    int oldstate;
    int newstate;
    __u16 family;
    __u16 sport;
    __u16 dport;
    unsigned char task[TASK_COMM_LEN];
};

const struct event *unused_4 __attribute__((unused));

// 保存需要过滤的源端口
struct
{
    __uint(type, BPF_MAP_TYPE_HASH);
    __uint(max_entries, MAX_ENTRIES);
    __type(key, __u16);
    __type(value, __u16);
} sports SEC(".maps");

// 保存需要过滤的目标端口
struct
{
    __uint(type, BPF_MAP_TYPE_HASH);
    __uint(max_entries, MAX_ENTRIES);
    __type(key, __u16);
    __type(value, __u16);
} dports SEC(".maps");

// 保存sock的最后一次变动时间
struct
{
    __uint(type, BPF_MAP_TYPE_HASH);
    __uint(max_entries, MAX_ENTRIES);
    __type(key, struct sock *);
    __type(value, __u64);
} timestamps SEC(".maps");

struct
{
    __uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
    __uint(key_size, sizeof(__u32));
    __uint(value_size, sizeof(__u32));
} events SEC(".maps");

const volatile __u16 target_family = 0;
const volatile bool filter_by_sport = false;
const volatile bool filter_by_dport = false;

SEC("tp/sock/inet_sock_set_state")
int handle_set_state(struct trace_event_raw_inet_sock_set_state *ctx)
{

    __u16 sport = ctx->sport;
    __u16 dport = ctx->dport;
    __u16 family = ctx->family;

    struct sock *sk = (struct sock *)ctx->skaddr;

    // 过滤 tcp 协议
    if (ctx->protocol != IPPROTO_TCP)
        return 0;

    // 判断是否指定过滤 ipv4 或者 ipv6
    if (target_family && target_family != family)
        return 0;

    // 判断是否过滤源端口
    if (filter_by_sport && !bpf_map_lookup_elem(&sports, &sport))
        return 0;

    // 判断是否过滤目标端口
    if (filter_by_dport && !bpf_map_lookup_elem(&dports, &dport))
        return 0;

    // 计算间隔时间
    __u64 *tsp, ts, delta_us;
    tsp = bpf_map_lookup_elem(&timestamps, &sk);
    ts = bpf_ktime_get_ns();
    if (!tsp)
    {
        delta_us = 0;
    }
    else
    {
        delta_us = (ts - *tsp) / 1000;
    }
    struct event event = {};

    event.skaddr = (__u64)sk;
    event.ts_us = ts / 1000;
    event.delta_us = delta_us;
    event.pid = bpf_get_current_pid_tgid() >> 32;
    event.oldstate = ctx->oldstate;
    event.newstate = ctx->newstate;
    event.family = family;
    event.sport = sport;
    event.dport = dport;
    bpf_get_current_comm(&event.task, sizeof(event.task));

    if (family == AF_INET)
    {
        // ipv4
        bpf_probe_read_kernel(&event.saddr, sizeof(event.saddr), &sk->__sk_common.skc_rcv_saddr);
        bpf_probe_read_kernel(&event.daddr, sizeof(event.daddr), &sk->__sk_common.skc_daddr);
    }
    else
    {
        // ipv6
        bpf_probe_read_kernel(&event.saddr, sizeof(event.saddr), &sk->__sk_common.skc_v6_rcv_saddr.in6_u.u6_addr32);
        bpf_probe_read_kernel(&event.daddr, sizeof(event.daddr), &sk->__sk_common.skc_v6_daddr.in6_u.u6_addr32);
    }

    bpf_perf_event_output(ctx, &events, BPF_F_CURRENT_CPU, &event, sizeof(event));

    if (ctx->newstate == TCP_CLOSE)
        bpf_map_delete_elem(&timestamps, &sk);
    else
        bpf_map_update_elem(&timestamps, &sk, &ts, BPF_ANY);

    return 0;
}

char LICENSE[] SEC("licenses") = "GPL";

File: cmd/14-tcpstate/tcpstate.go
package main

import (
	"bpf-developer-tutorial/pkg/utils"
	"bytes"
	"encoding/binary"
	"errors"
	"fmt"
	"github.com/cilium/ebpf"
	"github.com/cilium/ebpf/link"
	"github.com/cilium/ebpf/perf"
	"github.com/cilium/ebpf/rlimit"
	"github.com/spf13/cobra"
	"golang.org/x/sys/unix"
	"net"
	"os"
	"strings"
)

func runTcpState(cmd *cobra.Command, args []string) {

	err := rlimit.RemoveMemlock()
	if err != nil {
		panic(err)
	}

	spec, err := loadTcpstate()
	if err != nil {
		panic(err)
	}

	// 设置需要过滤的协议族
	var family uint16
	switch strings.ToLower(targetFamily) {
	case "ipv4":
		family = AfInet
	case "ipv6":
		family = AfInet6
	default:
		fmt.Printf("Invalid target family: %s. Using IPv4.\n", targetFamily)
		family = AfInet
	}

	err = spec.Variables["target_family"].Set(family)
	if err != nil {
		panic(err)
	}

	err = spec.Variables["filter_by_sport"].Set(filterBySport)
	if err != nil {
		panic(err)
	}

	err = spec.Variables["filter_by_dport"].Set(filterByDport)
	if err != nil {
		panic(err)
	}

	var obj tcpstateObjects
	err = spec.LoadAndAssign(&obj, nil)
	if err != nil {
		panic(err)
	}
	defer obj.Close()

	// 设置需要过滤的源端口
	err = setFilterPorts(obj.Sports, sportsList)
	if err != nil {
		panic(err)
	}

	// 设置需要过滤的目标端口
	err = setFilterPorts(obj.Dports, dportsList)
	if err != nil {
		panic(err)
	}

	tracepoint, err := link.Tracepoint("sock", "inet_sock_set_state", obj.HandleSetState, &link.TracepointOptions{})
	if err != nil {
		panic(err)
	}
	defer tracepoint.Close()

	reader, err := perf.NewReader(obj.Events, os.Getpagesize())
	if err != nil {
		panic(err)
	}
	defer reader.Close()

	go utils.ShutdownListenWithCallback(func() {
		reader.Close()
	})

	// 显示当前过滤设置
	fmt.Println("Current filter settings:")
	fmt.Printf("  IP Family: %s\n", targetFamily)
	fmt.Printf("  Filter by source ports: %v", filterBySport)
	if filterBySport {
		fmt.Printf(" (Ports: %s)", sportsList.String())
	}
	fmt.Println()
	fmt.Printf("  Filter by destination ports: %v", filterByDport)
	if filterByDport {
		fmt.Printf(" (Ports: %s)", dportsList.String())
	}
	fmt.Println()
	fmt.Println()

	fmt.Printf("%-16s %-8s %-16s %-16s %-8s %-10s %-10s %-12s -> %-12s %8s\n",
		"SKADDR", "C-PID", "C-COMM", "LADDR", "LPORT", "RADDR", "RPORT", "OLDSTATE", "NEWSTATE", "MS")

	var event StateEvent
	for {
		record, err := reader.Read()
		if err != nil {
			if errors.Is(err, perf.ErrClosed) {
				fmt.Printf("perf reader closed, exiting..")
				return
			}
			fmt.Printf("读取 perf 事件失败: %v", err)
			continue
		}

		err = binary.Read(bytes.NewReader(record.RawSample), binary.LittleEndian, &event)
		if err != nil {
			fmt.Printf("解析事件失败: %v", err)
			continue
		}

		processEvent(&event)

	}

}

func processEvent(event *StateEvent) {
	// 获取进程名
	comm := unix.ByteSliceToString(event.Task[:])

	// 格式化地址

	// 格式化状态名称
	oldState := getTcpStateName(event.Oldstate)
	newState := getTcpStateName(event.Newstate)

	// 打印事件信息
	fmt.Printf("%-16x %-8d %-16s %-15s %-8d %-15s %-8d %-12s -> %-12s %8d\n",
		event.Skaddr,
		event.Pid,
		comm,
		formatAddress(event.Saddr[:], event.Family),
		event.Sport,
		formatAddress(event.Daddr[:], event.Family),
		event.Dport,
		oldState,
		newState,
		event.DeltaUs/1000) // 转换为毫秒
}

// 根据地址族格式化 IP 地址
func formatAddress(addrBytes []byte, family uint16) string {
	switch family {
	case AfInet: // IPv4
		if len(addrBytes) >= 4 {
			// IPv4 地址通常存储在前 4 个字节
			addr := binary.LittleEndian.Uint32(addrBytes[:4])
			return formatIPv4(addr)
		}
		return "0.0.0.0"
	case AfInet6: // IPv6
		if len(addrBytes) >= 16 {
			return net.IP(addrBytes[:16]).String()
		}
		return "::"
	default:
		return "unknown"
	}
}

// 格式化 IPv4 地址
func formatIPv4(addr uint32) string {
	ip := make(net.IP, 4)
	binary.LittleEndian.PutUint32(ip, addr)
	return ip.String()
}

func setFilterPorts(m *ebpf.Map, ports []uint16) error {
	for _, port := range ports {
		err := m.Put(port, uint16(1))
		if err != nil {
			return err
		}
	}
	return nil
}
